src/java/org/apache/cassandra/db/tries/MemtableReadTrie.java

```diff
@@ -0,0 +1,921 @@
+     (not the block start) and can be thus negative since node pointers points towards the end of blocks.
+     */
+
+    // Limit for the starting cell / sublevel (2 bits -> 4 pointers).
+    static final int SPLIT_START_LEVEL_LIMIT = 4;
+    // Limit for the other sublevels (3 bits -> 8 pointers).
+    static final int SPLIT_OTHER_LEVEL_LIMIT = 8;
+    // Bitshift between levels.
+    static final int SPLIT_LEVEL_SHIFT = 3;
+
+    static final int SPARSE_CHILD_COUNT = 6;
+    // Offset to the first child pointer of a spare node (laid out from the start of the block)
+    static final int SPARSE_CHILDREN_OFFSET = 0 - SPARSE_OFFSET;
+    // Offset to the first transition byte of a sparse node (laid out after the child pointers)
+    static final int SPARSE_BYTES_OFFSET = SPARSE_CHILD_COUNT * 4 - SPARSE_OFFSET;
+    // Offset to the order word of a sparse node (laid out after the children (pointer + transition byte))
+    static final int SPARSE_ORDER_OFFSET = SPARSE_CHILD_COUNT * 5 - SPARSE_OFFSET;  // 0
+
+    // Offset of the flag byte in a prefix node. In shared blocks, this contains the offset of the next node.
+    static final int PREFIX_FLAGS_OFFSET = 4 - PREFIX_OFFSET;
+    // Offset of the content id
+    static final int PREFIX_CONTENT_OFFSET = 0 - PREFIX_OFFSET;
+    // Offset of the next pointer in a non-shared prefix node
+    static final int PREFIX_POINTER_OFFSET = LAST_POINTER_OFFSET - PREFIX_OFFSET;
+
+    // Initial capacity for the node data buffer.
+    static final int INITIAL_BUFFER_CAPACITY = 256;
+
+    /**
+     * Value used as null for node pointers.
+     * No node can use this address (we enforce this by not allowing chain nodes to grow to position 0).
+     * Do not change this as the code relies there being a NONE placed in all bytes of the block that are not set.
+     */
+    static final int NONE = 0;
+
+    volatile int root;
+
+    /*
+     EXPANDABLE DATA STORAGE
+
+     The tries will need more and more space in buffers and content lists as they grow. Instead of using ArrayList-like
+     reallocation with copying, which may be prohibitively expensive for large buffers, we use a sequence of
+     buffers/content arrays that double in size on every expansion.
+
+     For a given address x the index of the buffer can be found with the following calculation:
+        index_of_most_significant_set_bit(x / min_size + 1)
+     (relying on sum (2^i) for i in [0, n-1] == 2^n - 1) which can be performed quickly on modern hardware.
+
+     Finding the offset within the buffer is then
+        x + min - (min << buffer_index)
+
+     The allocated space starts 256 bytes for the buffer and 16 entries for the content list.
+
+     Note that a buffer is not allowed to split 32-byte blocks (code assumes same buffer can be used for all bytes
+     inside the block).
+
+     TODO: implement delay and retry on space hitting the 2GB barrier.
```

Comment:
nit: Should we throw up a Jira for this?

Reply:
Outdated TODO, removed.